cardano-api/src/Cardano/Api/Typed.hs

```diff
@@ -946,13 +985,130 @@ newtype Lovelace = Lovelace Integer
+data TxOutValue era where
+  -- | Transaction output amount value for a 'Byron'-era transaction.
+  ByronTxOutValue :: Byron.Lovelace -> TxOutValue Byron
+
+  -- | Transaction output amount value for a 'Shelley'-era transaction.
+  ShelleyTxOutValue :: Shelley.Coin -> TxOutValue Shelley
+
+  -- | Transaction output amount value for an 'Allegra'-era transaction.
+  AllegraTxOutValue :: Shelley.Coin -> TxOutValue Allegra
+
+  -- | Transaction output amount value for a 'Mary'-era transaction.
+  MaryTxOutValue
+    :: MultiAssetSupportedInEra Mary
+    -> Shelley.Value StandardMary
+    -> TxOutValue Mary
+
+deriving instance Eq (TxOutValue Byron)
+deriving instance Eq (TxOutValue Shelley)
+deriving instance Eq (TxOutValue Allegra)
+deriving instance Eq (TxOutValue Mary)
+deriving instance Show (TxOutValue Byron)
+deriving instance Show (TxOutValue Shelley)
+deriving instance Show (TxOutValue Allegra)
+deriving instance Show (TxOutValue Mary)
+
+
+-- | Representation of whether simple lovelace-only transactions are supported
+-- in a particular era.
+data SimpleTxOutValueSupportedInEra era where
+  -- | Simple lovelace-only transactions are supported in the 'Byron' era.
+  SimpleTxOutValueSupportedInByronEra
+    :: SimpleTxOutValueSupportedInEra Byron
+
+  -- | Simple lovelace-only transactions are supported in the 'Shelley' era.
+  SimpleTxOutValueSupportedInShelleyEra
+    :: SimpleTxOutValueSupportedInEra Shelley
+
+  -- | Simple lovelace-only transactions are supported in the 'Allegra' era.
+  SimpleTxOutValueSupportedInAllegraEra
+    :: SimpleTxOutValueSupportedInEra Allegra
+
+deriving instance Eq (SimpleTxOutValueSupportedInEra Byron)
+deriving instance Eq (SimpleTxOutValueSupportedInEra Shelley)
+deriving instance Eq (SimpleTxOutValueSupportedInEra Allegra)
+deriving instance Show (SimpleTxOutValueSupportedInEra Byron)
+deriving instance Show (SimpleTxOutValueSupportedInEra Shelley)
+deriving instance Show (SimpleTxOutValueSupportedInEra Allegra)
+
+
+-- | Representation of whether multi-asset transactions are supported in a
+-- particular era.
+data MultiAssetSupportedInEra era where
+  -- | Multi-asset transactions are supported in the 'Mary' era.
+  MultiAssetSupportedInMaryEra :: MultiAssetSupportedInEra Mary
+
+deriving instance Eq (MultiAssetSupportedInEra Mary)
+deriving instance Show (MultiAssetSupportedInEra Mary)
+
+
+-- | Policy identifier.
+newtype PolicyId = PolicyId (Hash Script)
+  deriving newtype (Eq, Ord)
+
+
+-- | Asset identifier.
+newtype AssetId = AssetId ByteString
+  deriving newtype (Eq, Ord)
+
+
+pattern SimpleTxOutValue
+  :: Lovelace
+  -> TxOutValue era
+pattern SimpleTxOutValue l <-
+  ( (\txOut ->
+      case txOut of
+        ByronTxOutValue l -> Just (Lovelace $ Byron.lovelaceToInteger l)
+        ShelleyTxOutValue (Shelley.Coin l) -> Just (Lovelace l)
+        AllegraTxOutValue (Shelley.Coin l) -> Just (Lovelace l)
+        _ -> Nothing
+    ) -> Just l
+  )
```

Comment:
Unidirectional because constructing a simple lovelace-only tx out value can fail in the Byron case.